app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+            ref_dance->timer_cancelled = false;
+            ref_dance->tap_dance_decided = false;
+            *tap_dance = ref_dance;
+            return 0;
+        }
+    }
+    // Unable to store new tap-dance in array active_tap_dances[].
+    return -ENOMEM;
+}
+
+// clear_tap_dance
+
+// Takes an active_tap_dance as an input and reset its position to
+// ZMK_BHV_TAP_DANCE_POSITION_FREE. Unecessary to update other
+// active_tap_dance variables because those with FREE positions will
+// be overwritten when a new tap dance is created
+static void clear_tap_dance(struct active_tap_dance *tap_dance) {
+    tap_dance->position = ZMK_BHV_TAP_DANCE_POSITION_FREE;
+}
+
+// stop_timer
+
+// Takes an active_tap_dance as an input before cancelling the
+// work associated with its release_timer
+// Returns timer_cancel_result
+static int stop_timer(struct active_tap_dance *tap_dance) {
+    int timer_cancel_result = k_delayed_work_cancel(&tap_dance->release_timer);
+    if (timer_cancel_result == -EINPROGRESS) {
+        // too late to cancel, we'll let the timer handler clear up.
+        tap_dance->timer_cancelled = true;
+    }
+    return timer_cancel_result;
+}
+
+// reset_timer
+
+// Takes an active_tap_dance and a zmk_behavior_binding_event
+// to reset its timer to trigger (tapping_term_ms) milliseconds
+// after the event
+static void reset_timer(struct active_tap_dance *tap_dance,
+                        struct zmk_behavior_binding_event event) {
+    // Start the timer if one hasn't started already. Reset it if one exists.
+    tap_dance->release_at = event.timestamp + tap_dance->config->tapping_term_ms;
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    LOG_DBG("%d tap-dance ms_left = %d", tap_dance->position, ms_left);
+    if (ms_left > 0) {
+        k_delayed_work_submit(&tap_dance->release_timer, K_MSEC(ms_left));
+    }
+}
+
+// press_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually pressed
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    // Press chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    return behavior_keymap_binding_pressed(&binding, event);
+}
+
+// release_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually released
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    // Release chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
```

Comment:
In what scenario is it possible for `counter` to become greater than `behavior_count`?  I think the rest of the code should make sure that `counter` can not exceed `behavior_count`, instead of defensively preventing that from occurring here.